msal/tests/Test.MSAL.NET.Integration/UnifiedCacheTests.cs

```diff
@@ -0,0 +1,541 @@
+
+            // make sure that adal v3 RT is usable by Msal
+            msalAuthResult = await msalPublicClient.AcquireTokenSilentAsync(MsalScopes, account);
+            ValidateMsalAuthResult();
+
+            // make sure Msal remove account api remove corresponding cache entities in all formats  
+            msalAccounts = await msalPublicClient.GetAccountsAsync();
+            Assert.AreEqual(1, msalAccounts.Count());
+            account = msalAccounts.First();
+
+            await msalPublicClient.RemoveAsync(account);
+
+            AssertAdalCacheIsEmpty();
+            AssertMsalCacheIsEmpty();
+        }
+
+        [TestMethod]
+        [TestCategory("UnifiedCache_IntegrationTests")]
+        public async Task UnifiedCache_Adalv3ToAdalV4ToMsal2MigrationIntegrationTest()
+        {
+            // acquire adal tokens using adalV4
+            adalAuthResult = await adal::Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContextIntegratedAuthExtensions.
+                AcquireTokenAsync(adalContext, AdalResource1, ClientId,
+                new adal::Microsoft.IdentityModel.Clients.ActiveDirectory.UserPasswordCredential(user.Upn, securePassword));
+            ValidateAdalAuthResult();
+
+            // simulate adalV3 token cache state by setting client info in adal cache entities to null 
+            // and clearing msal cache
+            UpdateAdalCacheSetClientInfoToNull();
+            ClearMsalCache();
+            AssertMsalCacheIsEmpty();
+
+            
+            // Migration to AdalV4 - acquire adal tokens using adalV4
+
+            // make sure that AT in AdalV3 format is used by AdalV4
+            Assert.AreEqual(1, adalCache.ReadItems().Count());
+            adalAuthResult = await adalContext.AcquireTokenSilentAsync(AdalResource1, ClientId);
+            Assert.AreEqual(1, adalCache.ReadItems().Count());
+
+            // acquire token to different resource
+            adalAuthResult = await adal::Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContextIntegratedAuthExtensions.
+                AcquireTokenAsync(adalContext, AdalResource2, ClientId,
+                new adal::Microsoft.IdentityModel.Clients.ActiveDirectory.UserPasswordCredential(user.Upn, securePassword));
+            ValidateAdalAuthResult();
+
+            // At this poing Adal cache contains RTs for the same account in diff format v3 and v4
+            Assert.IsTrue(adalCache.ReadItems().Count() == 2);
+
+            var msalAccounts = await msalPublicClient.GetAccountsAsync();
+            Assert.AreEqual(1, msalAccounts.Count());
+            var account = msalAccounts.First();
+            Assert.AreEqual(user.Upn, account.Username);
+            // make sure for the same account RT in V4 format preffered over V3 format
+            Assert.IsNotNull(account.HomeAccountId);
+            Assert.IsNotNull(account.Environment);
+
+            // validate than Adal writes only Rt in Msal format  an
```

Comment:
>format  an [](start = 57, length = 10)

this text update doesn't seem right... something is missing

Reply:
fixed